core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -14817,4 +15056,114 @@ private String getUserStoreDomainName(UserStoreManager userStoreManager) {
         }
         return domainNameProperty;
     }
+
+    /**
+     * Process and return the modifed claim values against claim URI. Add or remove specific claim values
+     * against old claim values of multi-valued claims. Replace old claim values from modified values for
+     * non multi-valued claims.
+     *
+     * @param oldClaimMap                      Map of claim URIs against old claim values of user.
+     * @param multiValuedClaimsToAdd           Map of multi-valued claim URIs against values need to be added to
+     *                                         old claim value.
+     * @param multiValuedClaimsToDelete        Map of multi-valued claim URIs against values need to be removed from
+     *                                         old claim value.
+     * @param claimsExcludingMultiValuedClaims Map of non multi-valued claim URIs against modified values to be stred.
+     * @return Map of claim URIs against the modified claim values.
+     */
+    private Map<String, String> getModifiedClaims(Map<String, List<String>> oldClaimMap,
+                                                  Map<String, List<String>> multiValuedClaimsToAdd,
+                                                  Map<String, List<String>> multiValuedClaimsToDelete,
+                                                  Map<String, List<String>> claimsExcludingMultiValuedClaims) {
+
+        Map<String, String> claims = new HashMap<>();
+        String separator = ",";
+        if (StringUtils.isNotEmpty(realmConfig.getUserStoreProperty(MULTI_ATTRIBUTE_SEPARATOR))) {
+            separator = realmConfig.getUserStoreProperty(MULTI_ATTRIBUTE_SEPARATOR);
+        }
+        if (claimsExcludingMultiValuedClaims != null) {
+            for (String claimURI : claimsExcludingMultiValuedClaims.keySet()) {
+                claims.put(claimURI,
+                        StringUtils.join(claimsExcludingMultiValuedClaims.get(claimURI).iterator(), separator));
+            }
+        }
+
+        // Get modified claim values for multi-valued claims.
+        if (multiValuedClaimsToAdd != null) {
+            for (String claimURI : multiValuedClaimsToAdd.keySet()) {
+                List<String> modifiedValue = new ArrayList<>();
+                if (oldClaimMap.containsKey(claimURI)) {
+                    modifiedValue.addAll(oldClaimMap.get(claimURI));
+                    modifiedValue.addAll(multiValuedClaimsToAdd.get(claimURI));
+                } else {
+                    modifiedValue.addAll(multiValuedClaimsToAdd.get(claimURI));
+                }
+                claims.put(claimURI, StringUtils.join(modifiedValue.iterator(), separator));
+            }
+        }
+        if (multiValuedClaimsToDelete != null) {
```

Comment:
Can we use MapUtils?

Reply:
A null check is sufficient here due to the following for loop.
wso2-support/carbon-kernel#984 (comment)